executor/new_executor_xapi.go

```diff
@@ -0,0 +1,420 @@
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/plan"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/types"
+	"github.com/pingcap/tidb/xapi"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+// NewXSelectIndexExec represents XAPI select index executor without result fields.
+type NewXSelectIndexExec struct {
+	tableInfo   *model.TableInfo
+	table       table.Table
+	asName      *model.CIStr
+	ctx         context.Context
+	supportDesc bool
+	isMemDB     bool
+	result      *xapi.SelectResult
+	subResult   *xapi.SubResult
+	where       *tipb.Expr
+
+	tasks      []*lookupTableTask
+	taskCursor int
+	indexOrder map[int64]int
+	indexPlan  *plan.PhysicalIndexScan
+
+	mu sync.Mutex
+}
+
+// Fields implements Exec Fields interface.
+func (e *NewXSelectIndexExec) Fields() []*ast.ResultField {
+	return nil
+}
+
+// Schema implements Exec Schema interface.
+func (e *NewXSelectIndexExec) Schema() expression.Schema {
+	return e.indexPlan.GetSchema()
+}
+
+// Close implements Exec Close interface.
+func (e *NewXSelectIndexExec) Close() error {
+	e.result = nil
+	e.subResult = nil
+	return nil
+}
+
+// Next implements Executor Next interface.
+func (e *NewXSelectIndexExec) Next() (*Row, error) {
+	if e.tasks == nil {
+		startTs := time.Now()
+		handles, err := e.fetchHandles()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		log.Debugf("[TIME_INDEX_SCAN] time: %v handles: %d", time.Now().Sub(startTs), len(handles))
+		e.buildTableTasks(handles)
+		limitCount := int64(-1)
+		if e.indexPlan.LimitCount != nil {
+			limitCount = *e.indexPlan.LimitCount
+		}
+		concurrency := 2
+		if limitCount == -1 {
+			concurrency = len(e.tasks)
+		} else if limitCount > int64(BaseLookupTableTaskSize) {
+			concurrency = int(limitCount/int64(BaseLookupTableTaskSize) + 1)
+		}
+		log.Debugf("[TIME_INDEX_TABLE_CONCURRENT_SCAN] start %d workers", concurrency)
+		e.runTableTasks(concurrency)
+	}
+	for {
+		if e.taskCursor >= len(e.tasks) {
+			return nil, nil
+		}
+		task := e.tasks[e.taskCursor]
+		if !task.done {
+			startTs := time.Now()
+			err := <-task.doneCh
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			task.done = true
+			log.Debugf("[TIME_INDEX_TABLE_SCAN] time: %v handles: %d", time.Now().Sub(startTs), len(task.handles))
+		}
+		if task.cursor < len(task.rows) {
+			row := task.rows[task.cursor]
+			task.cursor++
+			return row, nil
+		}
+		e.taskCursor++
+	}
+}
+
+// fetchHandle fetches all handles from the index.
+// This should be optimized to fetch handles in batch.
+func (e *NewXSelectIndexExec) fetchHandles() ([]int64, error) {
+	idxResult, err := e.doIndexRequest()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	handles, err := extractHandlesFromIndexResult(idxResult)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// Save the index order.
+	e.indexOrder = make(map[int64]int)
```

Comment:
If query do not need to keep result in order, we do not need to save handles' order.